plugins/quota/actor.go

```diff
@@ -0,0 +1,101 @@
+package quota
+
+import (
+	"os"
+	"path"
+
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/glusterd2/xlator"
+	"github.com/gluster/glusterd2/pkg/errors"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type quotadActor struct{}
+
+const (
+	quotaDaemonKey = "enable"
+)
+
+// isQuotadAffectedNode returns true if there are local bricks of volume on
+// which quota is enabled
+func isQuotadAffectedNode(volumes []*volume.Volinfo) bool {
+	for _, v := range volumes {
+		if !isQuotaEnabled(v) {
+			continue
+		} else if v.State != volume.VolStarted {
+			continue
+		} else {
+			bricks := v.GetLocalBricks()
+			if len(bricks) > 0 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (actor *quotadActor) Do(v *volume.Volinfo, key, value string, logger log.FieldLogger) error {
+	if key != quotaDaemonKey {
+		return nil
+	}
+	quotadDaemon, err := NewQuotad()
+	if err != nil {
+		return err
+	}
+	// Create pidfile dir if not exists
+	if err := os.MkdirAll(path.Dir(quotadDaemon.pidfilepath),
+		os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+
+	// Create logFiledir dir
+	if err := os.MkdirAll(path.Dir(quotadDaemon.logfilepath),
+		os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+
+	volumes, err := volume.GetVolumes()
+	if err != nil {
+		logger.WithError(err).Error("failed to get volumes")
+		return err
+	}
+	// Check for making changes to quotad on this node
+	// For both disable and enable we need to do the same (restart).
+	if !isQuotadAffectedNode(volumes) {
```

Comment:
I thought a little more about this code organisation. As mentioned earlier, you must rely on user inputs `key` and `value`. Your code is focusing on the end action to be taken i.e restarting quotad but ignores how you arrive at that decision and what data you'd use to arrive at that decision.

For better readability, this condition can be:

```go
if value == "off" && isQuotadStopRequired(volumes) {
    if err = daemon.Stop(quotadDaemon, true, logger); err != nil {
        ....
    }
}
```

Your code is relying on the fact that the `Volinfo` passed is the new/modified one which contains the user input. That makes understanding this code very hard. You shouldn't rely on contents of `Volinfo` but rather on the user input (key and value variables) even though `Volinfo` contains the same information.

You have re-ordered the txn steps such that updating `Volinfo` comes first. The purpose of that ordering is only to allow volgen to regenerate the volfiles and not to use rely on `Volinfo` here for decision making. In future, we may change the volgen to generate volfiles from custom/temporary `Volinfo`. The code written here should be resilient to that.

Reply:
done